Allow presentation without holder binding (#513)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

}
```

## Existing Parameters {#existing_parameters}

The following additional considerations are given for pre-existing Authorization Request parameters:

`nonce`:

> : REQUIRED. Defined in [@!OpenID.Core]. It is used to securely bind Verifiable Presentation(s) provided by the Wallet to the particular transaction. The Verifier MUST create a fresh, cryptographically random number with sufficient entropy for every Authorization Request, store it with its current session, and pass it in the `nonce` Authorization Request Parameter to the Wallet. See (#preventing-replay) for details. Values MUST only contain ASCII URL safe characters (uppercase and lowercase letters, decimal digits, hyphen, period, underscore, and tilde).

Author: To be discussed - I left it REQUIRED for now. Technically not required if only asking for presentations, but then again, it is not hard to always add a nonce.

> : REQUIRED. Defined in [@!OpenID.Core]. It is used to securely bind Verifiable Presentation(s) provided by the Wallet to the particular transaction. The Verifier MUST create a fresh, cryptographically random number with sufficient entropy for every Authorization Request, store it with its current session, and pass it in the `nonce` Authorization Request Parameter to the Wallet. See (#preventing-replay) for details. Values MUST only contain ASCII URL safe characters (uppercase and lowercase letters, decimal digits, hyphen, period, underscore, and tilde).

Reviewer: We could add similar text to the one in state (required if you are requesting anything that has a key binding)

> : REQUIRED. Defined in [@!OpenID.Core]. It is used to securely bind Verifiable Presentation(s) provided by the Wallet to the particular transaction. The Verifier MUST create a fresh, cryptographically random number with sufficient entropy for every Authorization Request, store it with its current session, and pass it in the `nonce` Authorization Request Parameter to the Wallet. See (#preventing-replay) for details. Values MUST only contain ASCII URL safe characters (uppercase and lowercase letters, decimal digits, hyphen, period, underscore, and tilde).

Reviewer 2: WG discussion, agreed to keep it required

`response_mode`:
: REQUIRED. Defined in [@!OAuth.Responses]. This parameter is used (through the new Response Mode `direct_post`) to ask the Wallet to send the response to the Verifier via an HTTPS connection (see (#response_mode_post) for more details). It is also used to request signing and encrypting (see (#jarm) for more details). 

`client_id`:
: REQUIRED. Defined in [@!RFC6749]. This specification defines additional requirements to enable the use of Client Identifier Prefixes as described in (#client_metadata_management).

`state`:
: REQUIRED under the conditions defined in (#nkb-credentials). Otherwise, `state` is OPTIONAL. `state` values MUST only contain ASCII URL safe characters (uppercase and lowercase letters, decimal digits, hyphen, period, underscore, and tilde).

## Requesting Presentations without Holder Binding Proofs {#nkb-credentials}

The primary use case of this specification is to request and present Verifiable
Presentations, i.e., Presentations that contain a Holder Binding proof.

However, there are use cases where the Verifier wants to request presentation of

> Credentials without a proof of Holder Binding. Examples for such use cases include

Reviewer 3: ```suggestion
Credentials without a proof of cryptographic Holder Binding. Examples for such use cases include
```

low-security Credentials that do not support Holder Binding (e.g., a cinema
ticket), Credentials that are bound to a biometric trait, or Credentials that
are bound to claims (e.g., a diploma). In some cases, Credentials may support
Holder Binding, but the Verifier may not require it for the Presentation.

A Verifier that requests and accepts a Presentation of a Credential without a
proof of Holder Binding accepts that the presented Credential may have been
replayed. (#preventing-replay) contains additional considerations for this case.

To request a Credential without proof of Holder Binding, the Verifier specifies a

> different format in the DCQL request as defined in (#dcql_query) and

Reviewer 3: would it make sense to add an example?

trusted on their own and this feature mainly aims to help data minimization by not
revealing information that would likely be rejected.

> `allow_replay`:
> : OPTIONAL. A boolean which indicates whether the Verifier requires a Holder Binding
> proof. The default value is `false`, i.e., a Verifiable Presentation with Holder Binding
> is requested. If set to `true`, the Verifier accepts a Credential without Holder Binding
> proof. The Wallet MUST NOT include a Holder Binding proof in this case.

Reviewer 2: ```suggestion
`require_cryptographic_holder_binding`:
: OPTIONAL. A boolean which indicates whether the Verifier requires a Cryptographic Holder Binding
proof. The default value is `true`, i.e., a Verifiable Presentation with Cryptographic Holder Binding
is required. If set to `false`, the Verifier accepts a Credential without Cryptographic Holder Binding
proof.
```

trusted on their own and this feature mainly aims to help data minimization by not
revealing information that would likely be rejected.

> `allow_replay`:

Reviewer 3: sounds weird, no one is interested in allowing replay - I would rather call this element "key_binding_required"

Author: Making this sound slightly weird and not like something you want to use was exactly the intention. I fear that Verifiers may just use it because "I don't need this weird proof thing anyway".

Reviewer 2: renamed to `require_cryptographic_holder_binding`

: OPTIONAL. A boolean which indicates whether the Verifier requires a Holder Binding
proof. The default value is `false`, i.e., a Verifiable Presentation with Holder Binding
is requested. If set to `true`, the Verifier accepts a Credential without Holder Binding
proof. The Wallet MUST NOT include a Holder Binding proof in this case.

> Depending on the Credential format, a different Credential Format Identifier may be required to request Presentations without Holder Binding. Details are defined in (#format_specific_parameters). 

Author: That would work, but if we only use `ldp_vc`, for example, it means another breaking change (the format parameter is not only used in the DCQL query). If we only use `ldp_vp`, the identifier would not be correct in case no holder binding is requested.

> Depending on the Credential format, a different Credential Format Identifier may be required to request Presentations without Holder Binding. Details are defined in (#format_specific_parameters). 

Author: The identifiers are both used in `vp_formats_supported` to distinguish the algorithm for the VC and the one for the VP (see [here](https://github.com/openid/OpenID4VP/pull/500/files#diff-63c80bfdcc58989ea965cff9db251b78ceb99b68a7afeb03066f440535863bf1)). @TimoGlastra As you recently looked into this - do you have an idea how to solve this?

> Depending on the Credential format, a different Credential Format Identifier may be required to request Presentations without Holder Binding. Details are defined in (#format_specific_parameters). 

Reviewer 4: We're already make breaking changes to `vp_formats_supported`, so I don't see much problem in making more breaking changes.

I do think it's important to stay consistent with OpenID4VCI. The same format identifiers are used there (and making the same changes to OpenID4VCI might be a bigger concern now)

Some options I see:
1. Update to a new combined format identifier, e.g. `w3c_di`
  - this is better aligned with how W3C (JSON-LD) credentials are secured today (it's now called data integrity proofs rather than linked data proofs).
  - we merge the `vp_formats_supported` into one object under `w3c_di`, similar to e.g. `mso_mdoc` which has `issuer_signed_alg_values` and `device_signed_alg_values`. I didn't want to introduce this change as part of #500 as I didn't think it was worth the breaking change, but if we're already merging the format identifiers I think using a different identifier does not really anymore anymore.
  - the format identifier for dcql always becomes `w3c_di`.
2.  We keep using `ldp_vc`, since that is already used in the DCQL query
  - We still remove the `ldp_vp` format from the `vp_formats_supported` and merge it under one object.
  - SD-JWT also uses `dc+sd-jwt` (previously `vc+sd-jwt`) to refer to both the VC and VP, so one could argue that using `vc` in the format identifier is okay for both VC and VP. Although I think the concern here would be that `ldp_vc` used to have a different meaning.

One thing with merging the format identifiers is that with W3C there's a completely different structure returned whether a VC or VP is returned, so you'd have to look into the `type` value to see if `VerifiableCredential` or `VerifiablePresentation` is used (and check based on the query `allow_replay` value what it should be). This seems reasonable and simple to me.

DCQL with W3C support is new and wasn't really defined up until like a week ago, so there's not any adoption of DCQL + W3C yet. So the main thing we should be concerned about is the `vp_formats_supported` structure breaking, but that is already breaking in #500

> Depending on the Credential format, a different Credential Format Identifier may be required to request Presentations without Holder Binding. Details are defined in (#format_specific_parameters). 

Author: The second solution is what I had in mind - it seems to be the smaller change from what we have right now.

> Depending on the Credential format, a different Credential Format Identifier may be required to request Presentations without Holder Binding. Details are defined in (#format_specific_parameters). 

Author: I went with the second solution, as it seems to be @paulbastian's preferred solution as well, i.e., there are only `jwt_vc_json` and `ldp_vc` now. The text is changed accordingly, please adapt #500, thanks!

> Depending on the Credential format, a different Credential Format Identifier may be required to request Presentations without Holder Binding. Details are defined in (#format_specific_parameters). 

Reviewer 5: That's my preferred solution as well, as it makes things more consistent with OpenID4VCI.

Verifiers MUST validate the VP Token in the following manner:

1. Validate the format of the VP Token as defined in (#response-parameters).
1. Check the individual Presentations according to the specific Credential or Presentation format requested:

>    1. Validate the integrity and authenticity of the Presentation and Credential.

Reviewer 3: what about the state?

Author: `state` is not in the VP Token that is validated here. We don't have a section on general response validation yet, I think.

There is a clause in the text I added further above that `state` must be validated, if used.

   1. Validate that the returned Credential(s) meet all criteria defined in the query in the Authorization Request (e.g., Claims included in the presentation).

>    1. Validate that all Presentations contain a cryptographic proof of Holder Binding (i.e., that they are Verifiable Presentations), unless specifically requested otherwise.
>    1. For Verifiable Presentations, validate the Holder Binding. See (#preventing-replay) for the checks required to prevent replay of a Verifiable Presentation.

Author: This is in practice the same step. I'll change the wording a bit to clarify that.
```suggestion
   1. Validate that all Presentations contain a cryptographic proof of Holder Binding (i.e., that they are Verifiable Presentations), unless specifically requested otherwise.
   1. For Verifiable Presentations, validate the Holder Binding, including the checks required to present replay described in (#preventing-replay).
```

# Security Considerations {#security_considerations}

## Preventing Replay of Verifiable Presentations {#preventing-replay} 

An attacker could try to inject Presentations obtained from (for example) a previous Authorization Response into another Authorization Response, thus impersonating the End-User that originally presented the respective Verifiable Presentation. Holder Binding aims to prevent such attacks.

### Presentations without Holder Binding Proofs

By definition, Presentations without Holder Binding (see (#nkb-credentials)) do
not prove that the Holder possesses the private Holder Binding key belonging to
the Credential. A Verifier that consumes Presentations without Holder Binding
accepts the risk that the Holder may have obtained the Credential from a third

> party (e.g., by playing the role of a Verifier) and that the Holder may not be

Reviewer 3: I suggest to add considerations how the binding could be achieved, like claims binding to other credentials.

Author: From my experience with similar discussions in SD-JWT, this is a bit of a rathole and there are a lot of opinions on this topic :-) I suggest to file an issue and add text later.

Reviewer 3: https://github.com/openid/OpenID4VP/issues/533

`type_values`:
: OPTIONAL. An array of string arrays that specifies the types of Verifiable Credentials that the Verifier accepts to be used in the Presentation. Each of the top-level arrays specifies one alternative to match the type of the Verifiable Credential against. Each inner array specifies a set of types that must be present in the `type` property of the Verifiable Credential, regardless of order or the presence of additional types. 

For example, if the DCQL query contains `"type_values": [["ACred", "BCred"], ["CCred"]]`, then a Verifiable Credential with the type `["DCred", "BCred", "ACred"]` would match, but a Verifiable Credential with the type `["BCred"]` would not match.

#### Claims Matching

The `claims_path` parameter in the Credential Query as defined in (#credential_query) is used to specify the claims that the Verifier wants to receive in the Presentation. When used in the context of W3C Verifiable Credentials, the `claims_path` parameter always matches on the root of Verifiable Credential (not the Verifiable Presentation). Examples are shown in the following subsections.

### Verifier Metadata

> The `vp_formats` parameter of the Verifier metadata MUST have the keys `jwt_vp_json`, `jwt_vc_json`, `ldp_vp` or `ldp_vc` (according to the format, as defined in (#jwt_vc) and (#ldp_vc)), and for `jwt_vp_json` and `ldp_vp`, the value MUST be an object consisting of the following name/value pair:

Author: Done, see discussion above.

This section illustrates the presentation of a Credential conformant to [@VC_DATA] that is signed using JWS, and does not use JSON-LD.

The Credential Format Identifiers are `jwt_vc_json` to request a W3C Verifiable Credential and `jwt_vp_json` to request a Verifiable Presentation compliant to the [@VC_DATA] specification.

> If `jwt_vc_json` is used, the Credential is returned without Holder Binding. Therefore, the Verifier MUST set the `allow_replay` parameter in the Credential Query to `true` if it is using `jwt_vc_json`. Otherwise, `allow_replay` MUST be `false`. Wallets MUST reject requests that do not follow these rules.

Reviewer 4: I feel like `allow_replay` is more of a result of not requiring key binding. But I don't think it's the best name for a query parameter.

I'm not sure if there's an overarching term. I think in OID4VCI it's called key binding (so something like 'require_key_binding`) or more generic that a presentation is required (so `require_verifiable_presentation`).

Author: I don't think `require_key_binding` sounds serious enough :-) I'd prefer a wording that conveys that this is a critial parameter.

Reviewer 3: I agree with @TimoGlastra .

### Format Identifier

> The Credential Format Identifier is `dc+sd-jwt`. By default, an SD-JWT [@!I-D.ietf-oauth-selective-disclosure-jwt] with a Key Binding JWT (SD-JWT+KB) is returned as the Verifiable Presentation. If `allow_replay` is set to `true` in the Credential Query, an SD-JWT without the Key Binding JWT is returned.

Author: Agree with moving the text further above.

I think `allow_replay` must be set to `true` any time an SD-JWT is presented without key binding. No matter if it supports key binding or not.

> The Credential Format Identifier is `dc+sd-jwt`. By default, an SD-JWT [@!I-D.ietf-oauth-selective-disclosure-jwt] with a Key Binding JWT (SD-JWT+KB) is returned as the Verifiable Presentation. If `allow_replay` is set to `true` in the Credential Query, an SD-JWT without the Key Binding JWT is returned.

Author: Text was changed to reflect that.
